rp2040-hal/src/clocks/mod.rs

```diff
@@ -22,21 +51,113 @@ impl ShareableClocks {
     }
 }
 
+const XOSC_MHZ: u32 = 12_000_000_u32;
+
+fn make_div<S: TryInto<Hertz<u64>>, F: TryInto<Hertz<u64>>>(
+    src_freq: S,
+    freq: F,
+) -> Result<u32, ()> {
+    let src_freq = *src_freq.try_into().map_err(|_| ())?.integer();
+    let freq = *freq.try_into().map_err(|_| ())?.integer();
+    let div: u64 = (src_freq << 8).wrapping_div(freq);
+    Ok(div as u32)
+}
 /// Abstraction layer providing Clock Management.
 pub struct ClocksManager {
     clocks: CLOCKS,
     shared_clocks: ShareableClocks,
 }
 impl ClocksManager {
     /// Exchanges CLOCKS block against Self.
-    pub fn new(mut clocks_block: CLOCKS) -> Self {
+    pub fn new(mut clocks_block: CLOCKS, watchdog: &mut Watchdog) -> Self {
+        // Start tick in watchdog
+        watchdog.enable_tick_generation(XOSC_MHZ as u8);
+
+        // Disable resus that may be enabled from previous software
+        clocks_block.clk_sys_resus_ctrl.write_with_zero(|w| w);
+
         let shared_clocks = ShareableClocks::new(&mut clocks_block);
         ClocksManager {
             clocks: clocks_block,
             shared_clocks,
         }
     }
 
+    /// Initialize the clocks
+    pub fn init(
+        &self,
+        _: &CrystalOscillator<Stable>,
+        _: &PhaseLockedLoop<Locked, PLL_SYS>,
+        _: &PhaseLockedLoop<Locked, PLL_USB>,
+    ) {
+        // Configure clocks
+        // CLK_REF = XOSC (12MHz) / 1 = 12MHz
+        let mut ref_clock = self.ref_clock();
+        let div = make_div(12u32.MHz(), 12u32.MHz()).unwrap();
+        // If increasing divisor, set divisor before source.
+        if div > ref_clock.get_div() {
+            ref_clock.set_div(div);
+        }
+        ref_clock.set_xosc_src();
+        ref_clock.await_select(2);
+        ref_clock.set_div(div);
+
+        // CLK SYS = PLL SYS (125MHz) / 1 = 125MHz
+        let mut sys_clock = self.sys_clock();
+        let div = make_div(125u32.MHz(), 125u32.MHz()).unwrap();
+        // If increasing divisor, set divisor before source.
+        if div > sys_clock.get_div() {
+            sys_clock.set_div(div);
+        }
+        sys_clock.set_pll_sys_auxsrc();
```

Comment:
Probably should make sure we're not on the aux mux before switching this. I know I've managed to get one of my test boards to reset without it resetting the clocks.

Reply:
Yeah there should be a generic switch mux function that takes care of that like the C sdk has.. However Rust types wont let me do that easily...